File: algo_2021/w5.-traverse-divide-and-conquer/17.Subsets/17_subsets__iteration.py
class Solution:
    """
    @param nums: A set of numbers
    @return: A list of lists
    """
    def subsets(self, nums):
        # write your code here
        res = []

        nums.sort()
        self.dfs(nums, 0, [], res)
        return res
    
    def dfs(self, nums, index, subset, res):
        res.append(list(subset))
        print(res)
        
        for i in range(index, len(nums)):
            # [1] -> [1, 2]
            # go find all subsets that begin with [1, 2]
            subset.append(nums[i])
            self.dfs(nums, i + 1, subset, res)
            #[1, 2] => [1]
            subset.pop()
        
        

File: algo_2021/s6.-dfs-comb-perm/17.Subsets/17_subsets__iteration.java
public class Solution {
    /**
     * @param nums: A set of numbers
     * @return: A list of lists
     */
    public List<List<Integer>> subsets(int[] nums) {
        // write your code here
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        dfs(nums, 0, new ArrayList<Integer>(), res);
        return res;
    }

    private void dfs(int[] nums, int index, ArrayList<Integer> subset, List<List<Integer>> res) {
        res.add(new ArrayList<Integer>(subset));
        
        for (int i = index; i < nums.length; i++) {
            subset.add(nums[i]);
            dfs(nums, i + 1, subset, res);
            subset.remove(subset.size() - 1);
        }
    }
}